commit e0bea6adec1b3cf84bf2cf17f8fc69ea678b4af4
Author: Vitaliy Filippov <vitalif@yourcmc.ru>
Date:   Fri Nov 16 02:59:14 2012 +0400

    Patch: html-emails
    Type: feature
    
    Patch for EnotifDiff extension. Enables HTML email support and additional headers. CustIS Bug 25475.

diff --git a/includes/DefaultSettings.php b/includes/DefaultSettings.php
index e67323d..bf2427b 100644
--- a/includes/DefaultSettings.php
+++ b/includes/DefaultSettings.php
@@ -1019,6 +1019,11 @@ $wgEnableEmail = true;
 $wgEnableUserEmail = true;
 
 /**
+ * Default content-type for e-mail. Set to text/html for HTML e-mail.
+ */
+$wgEmailContentType = 'text/plain';
+
+/**
  * Set to true to put the sending user's email in a Reply-To header
  * instead of From. ($wgEmergencyContact will be used as From.)
  *
diff --git a/includes/UserMailer.php b/includes/UserMailer.php
index d0eb919..7b613a9 100644
--- a/includes/UserMailer.php
+++ b/includes/UserMailer.php
@@ -117,15 +117,20 @@ class UserMailer {
 	 * @param $subject String: email's subject.
 	 * @param $body String: email's text.
 	 * @param $replyto MailAddress: optional reply-to email (default: null).
-	 * @param $contentType String: optional custom Content-Type (default: text/plain; charset=UTF-8)
+	 * @param $contentType String: optional custom Content-Type (default: $wgEmailContentType; charset=UTF-8)
+	 * @param $additionalHeaders Array: optional custom additional header (default: empty)
 	 * @return Status object
 	 */
-	public static function send( $to, $from, $subject, $body, $replyto = null, $contentType = 'text/plain; charset=UTF-8') {
+	public static function send( $to, $from, $subject, $body, $replyto = null, $contentType = null, $additionalHeaders = array()) {
 		global $wgSMTP, $wgEnotifMaxRecips, $wgAdditionalMailParams;
+		global $wgEmailContentType;
 
 		if ( !is_array( $to ) ) {
 			$to = array( $to );
 		}
+		if ( is_null( $contentType ) ) {
+			$contentType = $wgEmailContentType.'; charset=UTF-8';
+		}
 
 		wfDebug( __METHOD__ . ': sending mail to ' . implode( ', ', $to ) . "\n" );
 
@@ -170,13 +175,14 @@ class UserMailer {
 			$headers['Subject'] = self::mimeBase64( $subject );
 			$headers['Date'] = date( 'r' );
 			$headers['MIME-Version'] = '1.0';
-			$headers['Content-type'] = ( is_null( $contentType ) ?
-				'text/plain; charset=UTF-8' : $contentType );
+			$headers['Content-type'] = $contentType;
 			$headers['Content-transfer-encoding'] = '8bit';
 			// @todo FIXME
 			$headers['Message-ID'] = "<$msgid@" . $wgSMTP['IDHost'] . '>';
 			$headers['X-Mailer'] = 'MediaWiki mailer';
 
+			$headers += $additionalHeaders;
+
 			wfSuppressWarnings();
 
 			// Create the mail object using the Mail::factory method
@@ -209,17 +215,22 @@ class UserMailer {
 			}
 
 			$headers = array(
-				"MIME-Version: 1.0",
-				"Content-type: $contentType",
-				"Content-Transfer-Encoding: 8bit",
-				"X-Mailer: MediaWiki mailer",
-				"From: " . $from->toString(),
+				"MIME-Version" => "1.0",
+				"Content-type" => $contentType,
+				"Content-Transfer-Encoding" => "8bit",
+				"X-Mailer" => "MediaWiki mailer",
+				"From" => $from->toString(),
 			);
 			if ( $replyto ) {
-				$headers[] = "Reply-To: " . $replyto->toString();
+				$headers["Reply-To"] =  $replyto->toString();
 			}
 
-			$headers = implode( $endl, $headers );
+			$headers += $additionalHeaders;
+
+			$strHeaders = '';
+			foreach ( $headers as $key => $value ) {
+				$strHeaders .= "$key: $value$endl";
+			}
 
 			wfDebug( "Sending mail via internal mail() function\n" );
 
@@ -232,7 +243,7 @@ class UserMailer {
 				$to = array( $to );
 			}
 			foreach ( $to as $recip ) {
-				$sent = mail( $recip->toString(), self::mimeBase64( $subject ), $body, $headers, $wgAdditionalMailParams );
+				$sent = mail( $recip->toString(), self::mimeBase64( $subject ), $body, $strHeaders, $wgAdditionalMailParams );
 			}
 
 			restore_error_handler();
@@ -369,14 +380,17 @@ class EmailNotification {
 		$watchers = array();
 		if ( $wgEnotifWatchlist || $wgShowUpdatedMarker ) {
 			$dbw = wfGetDB( DB_MASTER );
-			$res = $dbw->select( array( 'watchlist' ),
+			$userCondition = array(
+				'user_id=wl_user',
+				'wl_title' => $title->getDBkey(),
+				'wl_namespace' => $title->getNamespace(),
+				'wl_user != ' . intval( $editor->getID() ),
+				'wl_notificationtimestamp IS NULL',
+			);
+			wfRunHooks('EnotifUserCondition', array(&$this, &$userCondition));
+			$res = $dbw->select( array( 'watchlist', 'user' ),
 				array( 'wl_user' ),
-				array(
-					'wl_title' => $title->getDBkey(),
-					'wl_namespace' => $title->getNamespace(),
-					'wl_user != ' . intval( $editor->getID() ),
-					'wl_notificationtimestamp IS NULL',
-				), __METHOD__
+				$userCondition, __METHOD__
 			);
 			foreach ( $res as $row ) {
 				$watchers[] = intval( $row->wl_user );
@@ -529,6 +543,7 @@ class EmailNotification {
 
 		$keys['$PAGETITLE'] = $this->title->getPrefixedText();
 		$keys['$PAGETITLE_URL'] = $this->title->getCanonicalUrl();
+		$keys['$PAGETITLE_URL_NOENC'] = urldecode( $this->title->getCanonicalUrl() );
 		$keys['$PAGEMINOREDIT'] = $this->minorEdit ? wfMsgForContent( 'minoredit' ) : '';
 		$keys['$PAGESUMMARY'] = $this->summary == '' ? ' - ' : $this->summary;
 		$keys['$UNWATCHURL'] = $this->title->getCanonicalUrl( 'action=unwatch' );
@@ -551,6 +566,8 @@ class EmailNotification {
 		$subject = strtr( $subject, $keys );
 		$this->subject = MessageCache::singleton()->transform( $subject, false, null, $this->title );
 
+		wfRunHooks('EnotifComposeCommonMailtext', array(&$this, &$keys));
+
 		$body = wfMsgExt( 'enotif_body', 'content' );
 		$body = strtr( $body, $keys );
 		$body = MessageCache::singleton()->transform( $body, false, null, $this->title );
@@ -638,6 +655,8 @@ class EmailNotification {
 				$wgContLang->time( $this->timestamp, true, false, $timecorrection ) ),
 			$body );
 
+		wfRunHooks( 'EnotifPersonalizeMailtext', array( &$this, &$watchingUser, &$body ) );
+
 		return UserMailer::send( $to, $this->from, $this->subject, $body, $this->replyto );
 	}
 
